src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -69,6 +69,97 @@ def __init__(
         self.upgrade_strategy = upgrade_strategy
         self._result: Optional[Result] = None
 
+    def _get_ireq(
+        self,
+        candidate: Candidate,
+        direct_url_requested: bool,
+    ) -> Optional[InstallRequirement]:
+        """Get the InstallRequirement to install for a candidate.
+
+        Returning None means the candidate is already satisfied by the current
+        environment state and does not need to be handled.
+        """
+        ireq = candidate.get_install_requirement()
+
+        # No ireq to install (e.g. extra-ed candidate). Skip.
+        if ireq is None:
+            return None
+
+        # The currently installed distribution of the same identifier.
+        installed_dist = self.factory.get_dist_to_uninstall(candidate)
+
+        if installed_dist is None:  # Not installed. Install incoming candidate.
+            return ireq
+
+        # If we return this ireq, it should trigger uninstallation of the
+        # existing distribution for reinstallation.
+        ireq.should_reinstall = True
+
+        # Reinstall if --force-reinstall is set.
+        if self.factory.force_reinstall:
+            return ireq
+
+        # The artifact represented by the incoming candidate.
+        cand_link = candidate.source_link
+
+        # The candidate does not point to an artifact. This means the resolver
+        # has already decided the installed distribution is good enough.
+        if cand_link is None:
+            return None
+
+        # The incoming candidate was produced only from version requirements.
+        # Reinstall if the installed distribution's version does not match.
```

Comment:
@uranusjr I've finally made some time to test this PR. So many combinations to test...

When we first install from a direct URL, then reinstall from a version only, shouldn't we reinstall always ? With this it does not reinstall if the version match.

Reply:
> So many combinations to test...

Very true, it’s difficult to keep everything stright in my head even when I wrote it, that’s why there are comments on literally every if branch.

> install from a direct URL, then reinstall from a version only

I’d expect this to not reinstall. With a version specifier, the user only cares about the version, so we only reinstall when the version does not match. If the direct URL requirement provides incompatible code, it should pick a different version number.

Reply:
But the direct URL is recorded, so if you pip freeze you get the direct URL and not the version requirement. So I think it would be more consistent to reinstall.

Reply:
> that’s why there are comments on literally every if branch.

Yes, they are very clear and helpful, thanks!

Reply:
> But the direct URL is recorded, so if you pip freeze you get the direct URL and not the version requirement.

It does but I don’t think it’s enough reason. This to me is similar to specifying `==2.0` and `>=1.0,<3.0`; the direct URL specifier is simply a stricter variant of the version specifier, and freeze outputting a direct URL instead of a version is IMO the correct output, since that URL is a valid way to satisfy the version specifier, and also the most descriptive way to express the user request.

Reply:
Ok, I see, it is a bit similar to when you have installed an old version of a package, and `pip install package` does not install the latest version because the installed version satisfies the request.

Reply:
On the other hand when we install from index over an editable install, it reinstalls, so there is some asymmetry here.

Also imaging this scenario:
1. install latest version of pkga from PyPI
2. to test the next version for which the author has not yet bumped the version number, install pkga from git
3. now if we want to come back from the release version, running `pip install pkga` would be my instinctive approach (and that would work if I had installed in editable mode in step 2)

Reply:
I want to back up this symmetry argument with a concrete use case.

Zulip has a list of [hundreds of transitively pinned requirements](https://github.com/zulip/zulip/blob/main/requirements/dev.txt), upgraded on a controlled schedule. Occasionally we need to transition a requirement from a PyPI version to a Git URL, to pull in a change that isn’t yet released on PyPI. We also need to be able to go backwards through our own history to bisect bugs, or to back out of a branch that’s being tested but not yet merged to `main`.

We don’t want to do a full `pip install --force-reinstall -r dev.txt` because that wastes a lot of time reinstalling hundreds of packages when only a few of them changed. So we want to be able to rely on `pip install -r dev.txt` to replace a PyPI release with a URL or vice versa, even if they have the same version number.

In response to the argument that “if the direct URL requirement provides incompatible code, it should pick a different version number”, I have two points: (1) in practice, actual upstream projects don’t bump their version number in Git for each individual incompatible change, and (2) I want to reproduce a set of *exact* requirements, not merely “compatible” ones.

(We’re currently stuck on pip 20.3.4 with `--use-deprecated=legacy-resolver` for this reason—this at least works in the PyPI → URL direction if we lie about the version number in the URL, but is of course terrible. That’s why I’ve been eagerly following this fix.)

Reply:
I wonder if it’s possible to overload `--upgrade` for this, i.e. `pip install -U foo` re-installs if the current `foo` is from a direct URL, but `pip install foo` does not. One complaint the old upgrade scheme in the legacy resolver is it reinstalls _too much_ and it’s been a thing I’d like to not do in the new design. BUt it’s pretty clear there’s no one-size solution for everyone. Using `--upgrade` for this would be quite consistent of other things this new logic uses the flag for.

Reply:
> I wonder if it’s possible to overload --upgrade for this

I'm not sure. I still plan to do more tests with `--upgrade` and try to understand what `--upgrade-strategy` actually does.

But for this particular case, I would find it unintuitive to have to use `pip install --upgrade -r requirements.txt` to achieve the expected result when `requirements.txt` comes out of pip freeze with exact requirements.